Spec/PushActivationStateMachine.swift

```diff
@@ -654,77 +660,772 @@ class PushActivationStateMachine : QuickSpec {
+
+                        waitUntil(timeout: testTimeout) { done in
+                            let simulatedError = NSError(domain: ARTAblyErrorDomain, code: 1234, userInfo: nil)
+                            let partialDone = AblyTests.splitDone(2, done: done)
+                            stateMachine.transitions = { event, from, to in
+                                if let event = event as? ARTPushActivationEventUpdatingRegistrationFailed {
+                                    expect(event.error.domain) == ARTAblyErrorDomain
+                                    expect(event.error.code) == simulatedError.code
+                                    stateMachine.transitions = nil
+                                    partialDone()
+                                }
+                            }
+                            delegate.onPushCustomRegister = { error, deviceDetails in
+                                expect(error).to(beNil())
+                                expect(deviceDetails).to(beIdenticalTo(rest.device))
+                                partialDone()
+                                return simulatedError
+                            }
+                            stateMachine.send(ARTPushActivationEventCalledActivate())
+                            expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateWaitingForRegistrationUpdate.self))
+                        }
+
+                        expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateAfterRegistrationUpdateFailed.self))
+                        expect(httpExecutor.requests.count) == 0
                     }
-                    defer { hookDevice.remove() }
 
-                    stateMachine.send(ARTPushActivationEventDeregistered())
-                    expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateNotActivated.self))
-                    expect(deactivatedCallbackCalled).to(beTrue())
-                    expect(setAndPersistIdentityTokenDetailsCalled).to(beTrue())
-                }
+                    it("should fire UpdatingRegistrationFailed event and include device auth") {
+                        expect(stateMachine.current).to(beAKindOf(ARTPushActivationStateAfterRegistrationUpdateFailed.self))
 
-                // RSH3g3
-                it("on Event DeregistrationFailed") {
-                    var deactivatedCallbackCalled = false
-                    let hook = stateMachine.testSuite_injectIntoMethod(after: NSSelectorFromString("callDeactivatedCallback:")) {
-                        deactivatedCallbackCalled = true
+                        let delegate = StateMachineDelegate()
+                        stateMachine.delegate = delegate
+
+                        let deviceIdentityToken = stateMachine.rest.device.identityTokenDetails?.token.base64Encoded()
+
+                        let simulatedError = NSError(domain: ARTAblyErrorDomain, code: 1234, userInfo: nil)
```

Comment:
This is repeated quite often. I suggest to have a `createError()` helper function.